sklearn/metrics/_pairwise_distances_reduction/__init__.py

```diff
@@ -0,0 +1,109 @@
+#
+#   Legend:
+#
+#      A ---⊳ B: A inherits from B
+#      A ---x B: A dispatches on B
+#      A ---* B: A composes B
+#
+#
+#                           (base dispatcher)
+#                       PairwiseDistancesReduction
+#                                   ∆
+#                                   |
+#                                   |
+#                 +-----------------+-----------------+
+#                 |                                   |
+#           (dispatcher)                        (dispatcher)
+#     PairwiseDistancesArgKmin           PairwiseDistancesRadiusNeighbors
+#           |                                              |
+#           |                                              |
+#           |                                              |
+#           |                  (64bit implem.)             |
+#           |          PairwiseDistancesReduction64        |
+#           |                       ∆                      |
+#           |                       |                      |
+#           |                       |                      |
+#           |     +-----------------+-----------------+    |
+#           |     |                                   |    |
+#           |     |                                   |    |
+#           x     |                                   |    x
+#    PairwiseDistancesArgKmin64       PairwiseDistancesRadiusNeighbors64
+#           |     ∆                                   ∆    |
+#           |     |                                   |    |
+#           x     |                                   |    |
+# FastEuclideanPairwiseDistancesArgKmin64             |    |
+#                 *                                   |    |
+#                 |                                   |    x
+#                 |            FastEuclideanPairwiseDistancesRadiusNeighbors64
+#                 |                                   *
+#                 |                                   |
+#                 +-----------------+-----------------+
+#                                   |
+#                            GEMMTermComputer64
+#
+#
+#    For instance :class:`PairwiseDistancesArgKmin`, dispatches to
+#    class:`PairwiseDistancesArgKmin64` if X and Y are both dense NumPy arrays
+#    with a float64 dtype.
+#
+#    In addition, if the metric parameter is set to "sqeuclidean",
+#    :class:`PairwiseDistancesArgKmin64` further dispatches to a subclass
+#    specialized to optimally handle the Euclidean distance case using the
+#    Generalized Matrix Multiplication (see the docstring of
+#    :class:`GEMMTermComputer64` for details).
+
+
+from ._base import _sqeuclidean_row_norms64
+
+from ._dispatcher import (
+    PairwiseDistancesReduction,
+    PairwiseDistancesArgKmin,
+    PairwiseDistancesRadiusNeighborhood,
+)
+
+__all__ = [
+    "PairwiseDistancesReduction",
+    "PairwiseDistancesArgKmin",
+    "PairwiseDistancesRadiusNeighborhood",
+    "_sqeuclidean_row_norms64",
```

Comment:
If `_sqeuclidean_row_norms64` is meant to be part of the Python API, maybe we should use fused-types to benefit from dispatching.

This way the abstraction of dtype-based specialized implementation would not leak into the Python-level API for this helper.

Reply:
Alternatively we could expose a Python level dispatching function for this helper in the `_dispatcher` submodule as we do for the classes.

This would allow to dispatch on memory layout (e.g. dense array vs sparse matrices).

Reply:
I like your second proposal.